### DISCUSSION TITLE
A Case for Web Components: Reconsidering React for Keycloak's GUIs

### DISCUSSION THREAD

**Discussion by stefanmaric** [keycloak/keycloak#24087] [Ideas]
# Context

Keycloak v22 removed the v1 of the Account Console theme.

Version 1 of the Account Console was a server-side, Multi-Page App (MPA) fully written in FreeMarker.

The successors, Version 2 and Version 3, of the Account Console are both React Single-Page Apps (SPA). Similarly, Version 2 of the Admin Console is also a React SPA.

I have a client I help maintain an instance of Keycloak, in order to upgrade to v22 I had to convert the custom Account Console I crafted for my client years ago into an SPA, as the `AccountProvider` on which the custom Service Provider Interface (SPI) & theme depended was removed.

I chose to rewrite it using the [CSP build of Alpine.js](https://alpinejs.dev/advanced/csp) because, while not a perfect solution, it doesn't need a build step, works with HTML template tags, its syntax doesn't interfere with Java FreeMarker, it's CSP-friendly (no use of `eval()`) and is relatively popular (charted in StackOverflow and StateOfJavascript developer surveys).

Before going with Alpine.js, I considered extending the React-based v2/v3 Account Console theme but it was a pointless task as not only would I have had to introduce a new runtime and workflow to the client's CI/CD pipeline but I would have had to rewrite most of the app because the client's design/theme uses a single view to display all aspects of the Account Client (that's it, user info, password, devices, MFA, clients, and everything else is rendered in a single page) while the base Account Console app is based on client-side routes.

I'm an avid and seasoned software consultant, I've been crafting and maintaining web apps for over 13 years, the last 7 of which were spent mostly using React, and while I love JSX as a templating language, appreciate the value of the React ecosystem, and hold dear my Functional Programming jargon, I don't believe React is the best option for an extensible product like Keycloak aims to be.

# Problems

* Extending React apps is (very) hard
  - You need to know React
  - Requires a build process, which in turn requires a Node.js setup
    * Because React relies on being a singleton (so that hooks and context can work)
  - Replacing React components is hard, because they cannot be extended (Class Components are not used, for good reasons)
    * So the only sensible way to extend a component is to copy its contents, complicates maintenance
    * Splitting the codebase into individual hook files might offer some extra granularity, but it doesn't solve the markup part
    * Likewise, extensibility could be improved with patters like Render Props, `as` polymorphic components, or `asChild`
    * This can be evidenced through [Keycloakify](https://www.keycloakify.dev/); a lot of effort has been put into such project and it still lacks an extension system, the workflow is still forking the repo and building on each change
    * Even though composability and interoperability are part of [React's design principles](https://legacy.reactjs.org/docs/design-principles.html), in practice, it is not meant to build pluggable systems, this might be explained by Meta's (formerly Facebook) problem space which dictates React's design decisions
* The React Core Team and Vercel (Next.js) are pushing hard for React Server Components
  - The future of React is as a server-side framework
  - You can get a taste of the future by looking at React's docs: https://react.dev/learn/start-a-new-react-project
  - None of the suggested options (Next.js, Remix, Gatsby, Expo) are SPAs
  - Which means that the future of React is tied to JavaScript server runtimes
  - Which means that Keycloak might be forced to be deployed as a dual-server cluster
  - Even though there's a renaissance of MPAs (which I support), SPAs are better for private apps (those behind auth)

# Proposal

---

**Build the next versions of the Account/Admin/Login Theme with a framework based on Web Components.**

---

I'm not the biggest fan of [Web Components](https://developer.mozilla.org/en-US/docs/Web/API/Web_components) (and less so of the attitude displayed by some of the most influential leaders of their community), but I have to admit that Web Components fits pretty well, better than anything else, the use-case of Keycloak.

## Which framework?

[**Lit**](https://lit.dev/) is the obvious choice because:

* It is popular.
* Backed by Google.
* `LitElement` is a lightweight abstraction on top of `CustomElement`.
* Doesn't require a build process.
* Fairly [small Javascript payload](https://bundlejs.com/?q=react%2Creact-dom&treeshake=%5B*%5D%2C%5B*%5D&text=%22console.log%28React%2CReactDOM%29%22) (not the smallest, but [React](https://bundlejs.com/?q=react%2Creact-dom&treeshake=%5B*%5D%2C%5B*%5D&text=%22console.log%28React%2CReactDOM%29%22) is almost 10 times bigger).
* Adopted by other big projects/companies, examples:
  - Adobe's Photoshop Web: https://medium.com/@addyosmani/photoshop-is-now-on-the-web-38d70954365a
  - Microsoft App Store: https://twitter.com/JudahGabriel/status/1710075338326454390

## Why not `x`?

There are many other (not as popular) options out there, why not those?

- [Stencil.js](https://stenciljs.com/): it requires a build process.
- [OMI](https://omijs.github.io/home/): likewise, requires a build process.
- [hibrids.js](https://hybrids.js.org/#/): too opinionated (includes routing and state management) and hides classes behind factory functions (Java developers might find the use of inheritance for building their UIs familiar or advantageous).
- Vanilla Custom Elements: lacking a templating language, it is too verbose and free-form for medium size apps like Keycloak's consoles.

# Reasoning

## Web Components are stable

They have been [supported in all major browsers](https://developer.mozilla.org/en-US/docs/Web/API/Web_components#browser_compatibility) since a while ago.

## The Platform™ will outlive React

Ignoring the debacle of the V0 spec of Web Components (namely Shadow DOM v0, Custom Elements v0, and HTML Imports), Web Components is very likely to be supported by all major browsers for decades to come, as browsers have a very good track record maintaining backwards compatibility that can be traced back to the [first ever published website](https://info.cern.ch/hypertext/WWW/TheProject.html) and the iconic [Space Jam hub](https://www.spacejam.com/1996/).

## Web Components are easy to extend, they are meant to

[Custom Elements](https://web.dev/articles/custom-elements-v1) are designed to be extensible and for developers to be able to extend browser's built-in elements (ignore the fact that Safari doesn't implement this).

Some hypothetical implementations:

```javascript
// my-theme/components/kc-user-avatar.js

// Get the original file this file is replacing.
// Where to place these files and/or how to access them would need to be defined.
// It could it be a convention and relative paths leveraging the `<base />` tag.
import { KCUserAvatar } from '../../../keycloak/components/kc-user-avatar.js'
import { html } from '../../../keycloak/vendor/lit.js'

// Create a new Custom Element that extends one of Keycloak's.
class MyCustomAvatar extends KCUserAvatar {
  // Replace markup
  render() {
    return html`
      <div>some custom markup to replace the default avatar<div>
    `
  }
}

// The consumers of this file will get a slightly different version of this component.
export { MyCustomAvatar as KCUserAvatar }
```

```javascript
// my-theme/components/kc-logout-button.js

import { KCLogoutButton} from '../../../keycloak/components/kc-logout-button.js'

class MyCustomLogoutButton extends KCLogoutButton {
  // Surgically modify specific behaviours.
  handleLogout(...args) {
    const doit = window.confirm('Are you sure you want to leave?')
    if (doit) super.handleLogout(...args)
  }
}

// Alternatively, we can let each component in charge of registering themselves.
customElements.define('kc-logout-button', MyCustomLogoutButton)
```

```javascript
// my-theme/main.js

// Import the original theme.
import '../../keycloak/main.js'

// Alternatively, we can simply override whole set of components at the top level.

customElements.define('kc-device-row', MyCustomDeviceRow)
customElements.define('kc-change-password-modal', MyCustomChangePasswordModal)
```

## Web Components open up a future of rich plugin systems

Imagine packing pieces of UI in custom SPIs jars that could be effortlessly imported and rendered into the different consoles. Web Components excel over React (or any other JavaScript framework) since they don't need a build process or have to share the same core, the framework is the browser.

Plugin authors can package their UI, built with whatever technologies they like, into a JavaScript module exporting one or more Custom Elements that comply to a specific set of attributes and properties depending on their intended use. These components can blend with the system theme via CSS Custom Properties.

## #noBuild is not the goal

The industry has recently seen a pushback against build pipelines, mainly driven by DHH: https://world.hey.com/dhh/we-tried-that-didn-t-work-d9c42fe1

And while I agree with some of the #noBuild premises, I'm by no means suggesting Keycloak to build its GUIs without build tools; on the contrary, I would push Keycloak and plugin authors to use build tools because there's a clear benefit in tools like Vite and TypeScript that should be harnessed.

My proposal is to design the Keycloak's frontend in such a way that's easily extensible and customizable, presents a low barrier of entry, and integrates well at runtime; coincidentally, not needing to rebuild the stack is a big part of it.

Lit's template language can be [compiled ahead of time](https://github.com/lit/lit/tree/main/packages/labs/compiler), but unlike other JavaScript frontend frameworks, compiled and uncompiled components can coexist in the same app.

The source code of a future Keycloak console could potentially be written in TypeScript and compiled for distribution and theme/plugin authors can still import and extend it at runtime and mix compiled with uncompiled components of their own.

# Challenges and caveats

## PatternFly

For those not familiar with Keycloak's development: [PatternFly](https://www.patternfly.org/) is the  Design System and Component Library used by RedHat's Open Source projects.

This would be the biggest pain-point while re-implementing themes with Web Components. There's an official (but community driven) [Web Components implementation of PatternFly](https://github.com/patternfly/patternfly-elements) but it is nowhere as developed and well maintained as its [React counterpart](https://github.com/patternfly/patternfly-react). I would anticipate a more uphill battle when developing with Web Components due to this.

On the other hand, this could be seen as an opportunity to upstream improvements to PatterFly other RedHat projects could benefit from.

## It is a migration

All migrations are hard, migrating to Web Components wouldn't be different. On a positive note, since Web Components can be used within any JS framework, there's a path for an incremental migration.

This migration could be done on the side as a community effort in a similar way Keycloakify inspired the new versions of the Admin and Account consoles. I would love to contribute, but as a solo developer with Keycloak not being the core of my consulting services, I have neither the capacity nor the inclination to lead such a project.

# References and related topics

* Discussions and posts highlighting the limitations of the React-based Keycloak themes
  - https://github.com/keycloak/keycloak/issues/19969
  - https://keycloak.discourse.group/t/keycloak-v2-account-theme/15626/1
  - https://github.com/keycloak/keycloak/discussions/19508
  - https://groups.google.com/g/keycloak-dev/c/j0aEZTWh-Lw
  - https://github.com/keycloak/keycloak/discussions/23893
  - https://github.com/keycloak/keycloak/discussions/23633
* Best summary about Web Components I could find: https://eisenbergeffect.medium.com/about-web-components-7b2a3ed67a78
* A comparison between React and Lit: https://blog.logrocket.com/lit-vs-react-comparison-guide/

# Closing

Thanks for reading if you made it this far. Would love to hear the thoughts of core maintainers and users alike.

**Comment by ssilvert** [comment]
@stefanmaric Thanks for your well-reasoned thesis.

We actually have considered Web Components as part of our customization strategy, but we have not embraced the idea as of yet.  We are pretty tied to what the PatternFly team does and this might become more viable as PatternFly Web Components becomes more mature.

We have worked with Keycloakify to ensure that their account consoles can still run on new versions of Keycloak.  This was recently merged and it probably would have saved you the effort of rewriting your own console.  Sorry that the timing was bad for you:
https://github.com/keycloak/keycloak/pull/22317
https://github.com/keycloak/keycloak/issues/22318

We may do something similar for admin console, which would allow you to do basically anything you want.

An approach we are playing with right now is the ability to just define a new section and we render it for you.  Basically, you give us the name of the left menu item along with metadata about the attributes you want to render in the content area.  And you would also provide the CRUD endpoints for the attributes.  I believe this will satisfy a very large number of use cases and would require minimal coding.  We could also make the renderer pluggable if you want to do your own thing with layouts and such.

**Comment by jonkoops** [comment]
First off, thank you for this very well written topic and elaborate motivation, it's something we always like to see and strive to do ourselves as well.

We're well aware of the problems that React brings in terms of customization, and the fact that we are now forcing users to also understand yet another framework that is further away from the stack that they know (FreeMarker templates and Java). Like @ssilvert mentioned we're attempting to offer a more granular approach to customization without needing to reach to the low-level stack that drives the UI, but we're also aware that for some this might not cover all the use-cases.

The reason why we decided to go for React is because it is the officially supported library by the PatternFly team, and since we had a hard deadline to deliver the new Administration Console (EOL of AngularJS), we did not have a lot of time to consider alternatives. Likewise, since the Account Console (v2) was based on React already we decided to do a quick port for (v3) to bring it more inline with the rest of our stack.

These are historical reasons, but they do not have to mean that we have to continue this trend. I too am more of a fan of progressive enhancement, especially for our less JavaScript intensive applications such as the login and registration flows. These do not make sense to implement using React, from an extensibility and performance perspective. Aside from the fact that these applications are also the most customized UIs we have, going far beyond what is usually done to the Admin and Account consoles.

I'd argue all your points are valid, and that we should try to deliver a solution that is more extensible and familiar to our users. However I do not yet feel like Web Components are perhaps the right fit for the login and registration specifically, and I'd like to stick to plain HTML there as much as possible to avoid needing JavaScript execution to render basic elements such as buttons, forms, etc. I'd prefer we keep things in templates and server-side rendered as much as possible. 

It looks like the [Declarative Shadow DOM](https://developer.chrome.com/en/articles/declarative-shadow-dom/) proposal could help us there, but it's still [not widely supported](https://caniuse.com/declarative-shadow-dom). But perhaps progressive enhancement is an option here, requiring older browsers to wait for JS execution. Naturally, we can use Web Components for smaller interactive 'islands' in these applications, we already have PRs that are implementing such functionalities (see #23574).

We'll be upgrading the login an registration to PatternFly 5 relatively soon, so I'll be taking that as an opportunity to create an inventory of the things that can be improved (input welcome).  One thing the team agrees on is that we will not be using React for these existing UIs. But we'll likely need to make a bet on some framework/library, whatever that may be.

A bit of a brain-dump but we might consider the following technologies:

- [Lit](https://lit.dev/)
- [HTMX](https://htmx.org/)
- [Alpine.js](https://alpinejs.dev/)
- [Quarkus Qute](https://quarkus.io/guides/qute)
- [PatternFly Elements](https://patternflyelements.org/)
